File: LeetCode_Java_WinterBreak2014/removeDuplicates.java
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

public class removeDuplicates {
    public static class Solution {
        public void myswap(int [] a, int i, int j) {
            int tmp = a[i];
            a[i] = a[j];
            a[j] = tmp;
        }
        public int removeDuplicates(int[] a) {
            if (a.length == 0) return 0;
            int idx = 0;
            for (int i = 1; i < a.length; i++) {
                if (a[idx] != a[i])
                    a[++idx] = a[i];
            }
            return idx + 1;
        }
        
        public int removeDuplicates0(int[] a) {
            if (a == null || a.length == 0) return 0;
            else if (a.length == 1) return 1;
            else if (a.length == 2) return a[0] == a[1] ? 1 : 2;
            int i = 0;
            int j = i + 1;
            int n = a.length;
            while (j < n) {
                while (j < n && a[j] == a[i]) j++;
                if (j == n) return i + 1;
                a[++i] = a[j++];
                while (j < n && a[j] != a[j - 1]) {
                    a[++i] = a[j++];
                }
            }
            return i + 1;
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        int [] a = {0, 0, 0, 0, 0, 1, 2, 2, 3, 3, 4, 4};
        System.out.println("");
        for(int i : a)
            System.out.print(i + ", ");
        System.out.println("");
        
        int res = result.removeDuplicates(a);
        System.out.println(res);
        for(int i : a)
            System.out.print(i + ", ");
        System.out.println("");
    }
}
/*
  Input:	[0,0,0,0,0,1,2,2,3,3,4,4]
  Output:	[0,0,0,0,1,2,2,3,4]
  Expected:	[0,1,2,3,4]

  Input:	[-3,-3,-2,-1,-1,0,0,0,0,0]
  Output:	[-3,0,-2,-1]
  Expected:	[-3,-2,-1,0]

  Input:	[-3,-1,-1,0,0,0,0,0,2]
  Output:	[-3,-1,2,0]
  Expected:	[-3,-1,0,2]

  Input:	[1,1,1,2]
  Output:	[1,2,1]
  Expected:	[1,2]

  Input:	[-1,0,0,0,0,3,3]
  Output:	[-1,0,3,0,3]
  Expected:	[-1,0,3]

  Input:	[1,1,2,3]
  Output:	[1,3,2]
  Expected:	[1,2,3]

  Input:	[1,2,2]
  Output:	[1,2,2]
  Expected:	[1,2]

  Runtime Error Message:	Line 4: java.lang.ArrayIndexOutOfBoundsException: -1
  Last executed input:	[1,1,1]
*/

File: myalgorithms/LeetCode_c++_0325-0430_2014-100/removeDuplicates.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int removeDuplicates(int a[], int n) {
    return distance(a, unique(a, a + n));
}

int removeDuplicates0(int a[], int n) {
    if (n == 0) return n;
    int cnt = 0; 
    bool flag = false;
    int i;
    
    for ( i = 1; i < n; ++i) {
        while (a[i] == a[cnt]) {
            if (i < n-1) 
                ++i;
            else  {
                flag = true;
                break;
            }
        }
        if (!flag)
            a[++cnt] = a[i];
    }

    if (i < n && a[i] == a[cnt])
        return cnt;
    else
        return cnt+1;    
}

int main() {
    int m = 4;
    int a[] = {-3, -1, 0, 0};

    cout << "a: ";
    for (int i = 0; i < m; ++i)
        cout << a[i] << ", ";
    cout << endl;
    
    int result = removeDuplicates(a, m);
    cout << "result: " << result << endl;
    for (int i = 0; i < result; ++i)
        cout << a[i] << ", ";
    cout << endl;
    
    return 0;
}

/*
Input:	[-3,-1,0,0]
Output:	[-3,-1]
Expected:	[-3,-1,0]
*/
